src/action-providers.ts

```diff
@@ -21,9 +21,73 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+    t.traverseAST(ast, {
```

Comment:
I think that make sense indeed.

We already optimized this check a bit by parsing the code into AST once. It's only logical that we'd want to traverse the AST once and gather the feedback from the different refactorings 👍 

I agree with the direction of the PR.

From experience, what could go wrong would be one refactoring having side-effect on the AST traversal (e.g. stopping the traversal or modifying the traversed tree). I think it should be fine because most checks just set a boolean, but I'm not 100% confident on this.

Reply:
I agree. It should be documented that .stop() is not allowed in such refactoring implementations.